[campaignion_tracking] add Google Analytics 4

katzmo opened the pull request.

katzmo committed:
[tracking] JS upgrade

Switch from parcel 1 to microbundle and upgrade dependencies.

katzmo committed:
[tracking] JS: fix linter errors

katzmo committed:
[tracking] JS: update yarn.lock file

katzmo committed:
[tracking] JS: fix global variable

Previously parcel --global exported the modules as global variables.
It looks like only the tracker is actually needed.

katzmo committed:
[tracking] add GA4 sub-module as copy of GTM

katzmo committed:
[tracking] GA4 recommended events for donations

katzmo committed:
[tracking] GA4 custom events for submissions and opt-ins

katzmo committed:
[tracking] add list of events to README

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] fire events for all multistep forms

- `draftBegin` on the 2nd step
- `draftContinue` from the 3rd step on

katzmo committed:
[tracking] push GA4 events more similar to gtag()

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] add snippets to submodules

When a tag ID is provided, GTM/GA4 will be initialized by the respective
sub-module.

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+/**
+ * Mapping of codes to eventNames.
+ *
+ * This also is a whitelist of which events should *also* be recognized as
```
This could easily say “list“ instead of “whitelist“ and also I think there is one more also than needed.

katzmo replied:
Not even sure the list is needed? :thinking:

katzmo replied:
Have removed it and replaced all occurrences of "whitelist".

torotil replied:
Thank you!

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+    }
+
+    if (typeof this.dataLayer === 'undefined') {
+      this.printDebug('No datalayer found. Doing nothing.')
```
Is there a return `return` missing after this statement?

katzmo replied:
Looks like it.

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+      event: 'checkoutBegin',
+      ecommerce: {
+        checkout: {
+          actionField: { step: 1 }, // begin == 1
```
The comment here is funny. The one thing it says is the only thing that’s obvious from the code.

Is this something required bei GA? Is this something we define arbitrarily (and thus might need user documentation)?

katzmo replied:
No idea, I've assumed it means something for GTM :shrug: It’s not part of the GA4 event.

torotil replied:
Ok, then never mind :grimacing:

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+  /**
+   * Handle "donationSuccess".
+   *
+   * The event data needs to include an transaction id.
+   * Falls back to use a random number.
```
This comment contradicts itself: Either it is required or there is a default. It can’t be both.

katzmo replied:
I don’t think so. The transaction id is required by google, not by this method. If no id is passed, the method will create a random id to send in the event.

torotil replied:
I don’t think it’s obvious that this comment refers to GA’s needs instead of the function’s parameters.

katzmo replied:
I didn't write the comment and I find it pretty clear. But you’re welcome to make a suggestion for improvement!

torotil replied:
Yep. The perils  of copy&pasting code from somewhere else and putting it to review again. I’ve been on the other side of this often enough. Sorry if it’s frustrating.

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_setDonationProduct (eventName, eventData, context) {
```
I think the logic of this function could be simplified if it would separate the product comparison from preparing the events. This also applies to the GA4 version of the next commit. So:

1. Build new product & compare it to the old, if equal then return.
2. Build events and send them.

torotil replied:
Also the product comparison seems like something that’s needed in several integrations. The base integration could take care of doing this and fire an event only when the product actually changes.

katzmo replied:
I’m afraid it's a bit more complicated. As much as I'd like to reduce unnecessary "setDonationProduct" events, this would need to keep track of the previous product data for comparison. Where/How would you suggest to do that? :thinking:

katzmo replied:
Turns out the current version of the GA4 code didn’t even work :innocent: 

Have refactored the product comparison logic and deleting products, choosing session storage as this was already used by other bits of codes.

torotil replied:
> Where/How would you suggest to do that?

I’m assuming that the product data is the same for all integrations (which I didn’t check thoroughly). Then campaignion_tracking could do the comparison and each individual integration could be called only when it actually changes.

torotil replied:
Ah I see that you already did refactor that. Awesome! :clap:

torotil commented on campaignion_tracking/js/campaignion_tracking.js:
```diff
@@ -8,7 +8,7 @@ Drupal.behaviors.campaignion_tracking.attach = function(context, settings) {
    */
 
-  function gracefulDispatch (name, data, trackingContext) {
+  function gracefulDispatch (type, name, data, trackingContext) {
```
This could just as well be defined outside the behavior which makes the behavior simpler and makes it clear it’s pure function (not a closure).

katzmo replied:
I think it makes it harder to find the Drupal behavior… :thinking:

torotil replied:
> I think it makes it harder to find the Drupal behavior…

It could also be below the behavior. Another method of the behaviour would also be an improvement.

torotil replied:
Much better. Thanks.

torotil commented on campaignion_tracking/js/campaignion_tracking.js:
```diff
@@ -19,28 +19,58 @@ Drupal.behaviors.campaignion_tracking.attach = function(context, settings) {
+    // Fire `draftBegin` after completing the first step of multi-step forms.
+    if (!settings['campaignion_tracking']['sent'].includes('draftBegin')) {
+      if (webform['last_completed_step'] === 1) {
+        settings['campaignion_tracking']['sent'].push('draftBegin');
```
Using Drupal.settings to store state is very weird it’s supposed to be only used for reading settings passed by the backend. Could `this` be used instead?

katzmo replied:
No, the data needs to be persistent between behavior calls. An alternative might be to use the session storage (this could also help with comparing products as discussed in a previous comment).

katzmo replied:
I think sessionStorage actually works quite well!

torotil replied:
`Drupal.behaviors.campaignion_tracking` has the same life time as `Drupal.settings`, so it would survive across AJAX calls but not across page loads.

Would the sessionStore need deletion when the donation is completed?

katzmo replied:
Hm, removing the storage data would be a good idea, otherwise the events are only triggered on the first form of a session. But when should we do that, how can the behavior know the form is complete or a different form than before? :thinking:

torotil replied:
The former code that stored the data in settings did that automatically: The settings are kept during AJAX calls but not for new actions, because that would mean a page load.

With the sessionStore you could still check for page loads in the behavior (context=window.document) and reset the storage. That has the advantage that in theory you could also track the transaction as “cancelled“.

katzmo replied:
Shouldn’t the tracking work without AJAX enabled on forms as well? (Up until now this would have just triggered events too often.)

katzmo replied:
I've pushed a new commit to store the data per form id instead of removing it. This should make the tracking possible for multiple forms in the same session, with and without AJAX enabled. 

When the user comes back to a previously tracked form in the same session one could argue this continues the submission from a tracking perspective so completing certain steps is not tracked again :innocent:

torotil replied:
What if someone does a one-time donation and a regular one with the same form? But yes, I’m not even sure what the behaviour should be.

katzmo replied:
It would sent ~~a "remove from cart" event for the one-time donation,~~ a "add to cart" event for the regular one and another "purchase" event, but not another "begin checkout" or "end checkout".

torotil replied:
This sounds like something that’s very hard to answer when somebody asks for it. Is there some documentation this would fit in?

katzmo replied:
I've added a new paragraph to the README.

torotil reviewed (changes requested):
I’ve found a few things that I think might be worth changing. I’m not sure how far we should push this as this is mostly a c&p of existing code (with potential of improvement). :innocent: Happy to discuss what’s worth the effort and what isn’t.

katzmo committed:
[tracking] reformat JS

katzmo committed:
[tracking] replace “whitelist”

katzmo committed:
[tracking] js fixes

- add missing return statements
- remove unused constants

katzmo committed:
[tracking] move data from settings to session storage

katzmo committed:
[tracking] only fire setDonationProduct on product change

Uses the original data for removed products. This is a breaking change
but makes the data more consistent and predictable.

Modules/themes implementing the JS hook might need to modify
`removeData` now as well.

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] store sent events per form

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -0,0 +1,426 @@
+   * Handle "donationSuccess".
+   *
+   * The event data needs to include an transaction id.
+   * Falls back to use a random number.
```
```suggestion
   * GA requires a transaction ID to be included in the event. If none is
   * passed this function generates a random ID.
```

katzmo committed:
Improve documentation

katzmo committed:
[tracking] GA4: use proper gtag() function

Pushing an array does not work, seems to require an arguments object.

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] print final data of all events in debug mode

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] fix missing product data

katzmo committed:
[tracking] don’t fire continueDraft when going back

katzmo committed:
[tracking] GA4: adapt custom events after feedback

katzmo committed:
[tracking] GTM: adapt custom events after feedback

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] document all events + parameters

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -73,6 +73,13 @@ export class GA4Tracker {
   }
 
+  /**
+   * Copy gtag function as required by GA4.
```
How about

> Generate an arguments object and pass it to the dataLayer.

This would tell us why the function is needed as well.

(github doesn’t provide me with the option to suggest an edit for this line.)

katzmo replied:
Have updated the comment. To me the most important info is still that it's a copy of the gtag function.

torotil commented on campaignion_tracking/package/src/ga4.js:
```diff
@@ -451,11 +452,11 @@ export class GA4Tracker {
+          item_id: product.id,
+          price: product.price,
+          item_variant: product.variant,
+          quantity: product.quantity,
```
This commit shows nicely that the generation of the product data from `eventData` and `this._context` could be done in one place and called from all 3 handlers. What do you think?

katzmo replied:
It could. Also makes more sense to store the formatted data than the raw one. Have pushed a new commit.

torotil reviewed (changes requested):
Thanks for the improvements! I have two more suggestions :innocent:

katzmo committed:
[tracking] GA4: update comment for gtag()

katzmo committed:
[tracking] GA4: format product data in 1 place

katzmo committed:
[tracking] recompile JS

katzmo committed:
[tracking] fix optin event data

katzmo committed:
[tracking] recompile JS
